packages/native-stack/src/views/NativeStackView.tsx

```diff
@@ -86,6 +118,145 @@ const MaybeNestedStack = ({
   return content;
 };
 
+type NavigationRoute<
+  ParamList extends ParamListBase,
+  RouteName extends keyof ParamList
+> = Route<Extract<RouteName, string>, ParamList[RouteName]> & {
+  state?: NavigationState | PartialState<NavigationState>;
+};
+
+const RouteView = ({
+  descriptors,
+  route,
+  index,
+  navigation,
+  state,
+  setNextDismissedKey,
+}: {
+  descriptors: NativeStackDescriptorMap;
+  route: NavigationRoute<ParamListBase, string>;
+  index: number;
+  navigation: NativeStackNavigationHelpers;
+  state: StackNavigationState<ParamListBase>;
+  setNextDismissedKey: React.Dispatch<React.SetStateAction<string | null>>;
+}) => {
+  const { options, render: renderScene } = descriptors[route.key];
+  const {
+    gestureEnabled,
+    headerShown,
+    animationTypeForReplace = 'pop',
+    animation,
+    orientation,
+    statusBarAnimation,
+    statusBarHidden,
+    statusBarStyle,
+  } = options;
+
+  let { presentation = 'card' } = options;
+
+  if (index === 0) {
+    // first screen should always be treated as `card`, it resolves problems with no header animation
+    // for navigator with first screen as `modal` and the next as `card`
+    presentation = 'card';
+  }
+
+  const isHeaderInPush = isAndroid
+    ? headerShown
+    : presentation === 'card' && headerShown !== false;
+
+  const isParentHeaderShown = React.useContext(HeaderShownContext);
+  const insets = useSafeAreaInsets();
+  const headerStatusBarHeight = isParentHeaderShown ? 0 : insets.top;
```

Comment:
Is this correct in case of native stack? e.g. if it's nested inside of another stack (JS or native), will it hide the statusbar inset?